optuna/samplers/_nsga2/crossover.py

```diff
@@ -0,0 +1,397 @@
+
+    x_min = np.min(xs, axis=0)
+    x_max = np.max(xs, axis=0)
+
+    x_diff = np.clip(x_max - x_min, 1e-10, None)
+    beta1 = 1 + 2 * (x_min - xl) / x_diff
+    beta2 = 1 + 2 * (xu - x_max) / x_diff
+    alpha1 = 2 - np.power(beta1, -(eta + 1))
+    alpha2 = 2 - np.power(beta2, -(eta + 1))
+
+    r = rng.uniform(0, 1, size=len(xs[0]))
+    mask1 = r > 1 / alpha1
+    betaq1 = np.power(r * alpha1, 1 / (eta + 1))
+    betaq1[mask1] = np.power((1 / (2 - r * alpha1)), 1 / (eta + 1))[mask1]
+
+    mask2 = r > 1 / alpha2
+    betaq2 = np.power(r * alpha2, 1 / (eta + 1))
+    betaq2[mask2] = np.power((1 / (2 - r * alpha2)), 1 / (eta + 1))[mask2]
+
+    c1 = 0.5 * ((x_min + x_max) - betaq1 * x_diff)
+    c2 = 0.5 * ((x_min + x_max) + betaq2 * x_diff)
+
+    v = []
+    for c1_i, c2_i, x1_i, x2_i in zip(c1, c2, xs[0], xs[1]):
+        if rng.rand() < 0.5:
+            if rng.rand() < 0.5:
+                v.append(c1_i)
+            else:
+                v.append(c2_i)
+        else:
+            if rng.rand() < 0.5:
+                v.append(x1_i)
+            else:
+                v.append(x2_i)
+    return np.array(v)
+
+
+def _vsbx(
+    xs: np.ndarray,
+    rng: np.random.RandomState,
+    eta: float,
+) -> np.ndarray:
+    r = rng.uniform(0, 1, size=len(xs[0]))
+    x0, x1 = xs[0], xs[1]
+    beta_1 = np.power(1 / 2 * r, 1 / (eta + 1))
+    beta_2 = np.power(1 / 2 * (1 - r), 1 / (eta + 1))
+    mask = r > 0.5
+    c1 = 0.5 * ((1 + beta_1) * x0 + (1 - beta_1) * x1)
+    c1[mask] = 0.5 * ((1 - beta_1) * x0 + (1 + beta_1) * x1)[mask]
+    c2 = 0.5 * ((3 - beta_2) * x0 - (1 - beta_2) * x1)
+    c2[mask] = 0.5 * (-(1 - beta_2) * x0 + (3 - beta_2) * x1)[mask]
+
+    v = []
+    for c1_i, c2_i, x1_i, x2_i in zip(c1, c2, x0, x1):
+        if rng.rand() < 0.5:
+            if rng.rand() < 0.5:
+                v.append(c1_i)
+            else:
+                v.append(c2_i)
+        else:
+            if rng.rand() < 0.5:
+                v.append(x1_i)
+            else:
+                v.append(x2_i)
+    return np.array(v)
+
+
+def _undx(
+    xs: np.ndarray, rng: np.random.RandomState, sigma_xi: float, sigma_eta: float
+) -> np.ndarray:
+    # https://www.jstage.jst.go.jp/article/sicetr1965/36/10/36_10_875/_pdf
+    x0, x1, x2 = xs[0], xs[1], xs[2]
+    n = len(x0)
+    xp = (x0 + x1) / 2
+    d = x0 - x1
+    D = _distance_from_x_to_PSL(x0, x1, x2)
+    xi = rng.normal(0, sigma_xi ** 2)
+    etas = rng.normal(0, sigma_eta, size=n)
+    es = _orthonormal_basis_vector_to_PSL(x0, x1)
+    one = xp
+    two = xi * d
+    three = np.zeros(len(es[0]))
+    for i in range(n - 1):
+        three += etas[i] * es[i]
+    three *= D
+    return one + two + three
+
+
+def _normalized_x1_to_x2(x1: np.ndarray, x2: np.ndarray) -> np.ndarray:
+    v_12 = x2 - x1
+    m_12 = np.linalg.norm(v_12, ord=2)
+    e_12 = v_12 / np.clip(m_12, 1e-10, None)
+    return e_12
+
+
+def _distance_from_x_to_PSL(x1: np.ndarray, x2: np.ndarray, x3: np.ndarray) -> np.ndarray:
```

Comment:
For lower snake case convention.
```suggestion
def _distance_from_x_to_psl(x1: np.ndarray, x2: np.ndarray, x3: np.ndarray) -> np.ndarray:
```